### PR TITLE
NIAD-910: Generate UUID for specimen role id element.

### PR DESCRIPTION
**Pull request by matthewdoreilly**
## Description
* Generate UUID for specimen role id element (Was mistakingly a hardcoded UUID)
* Minor refactoring
    * Move unit test data for specimens and observations under diagnosticreport level
    * Move code execution order in SpecimenMapper and ObservationMapper for readability
* Add unit test coverage for mapping interpretation -> interpretationCode for Observations

## Jira Ticket
https://gpitbjss.atlassian.net/browse/NIAD-910

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java [hunk 1]
```diff
@@ -215,24 +161,85 @@ private String mapObservationToObservationStatement(IdMapper idMapper, Observati
...
         }
 
-        return StringUtils.EMPTY;
+        CodeableConceptMappingUtils.extractTextOrCoding(observation.getDataAbsentReason()).ifPresent(dataAbsentReason -> {
+            String comment = DATA_ABSENT_PREFIX + dataAbsentReason;
+
+            narrativeStatementsBlock.append(
+                prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper)
+            );
+        });
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java:204]:
```suggestion
        CodeableConceptMappingUtils.extractTextOrCoding(observation.getDataAbsentReason())
            .map(DATA_ABSENT_PREFIX::concat)
            .map(comment -> prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper))
            .ifPresent(narrativeStatementsBlock::append);
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java [hunk 2]
```diff
@@ -66,40 +66,30 @@
     private final RandomIdGeneratorService randomIdGeneratorService;
```
Comment (matthewdoreilly) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java:66]: No functional changes here. Minor refactoring changes for readability

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java [hunk 3]
```diff
@@ -215,24 +161,85 @@ private String mapObservationToObservationStatement(IdMapper idMapper, Observati
...
+            narrativeStatementsBlock.append(
+                prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper)
+            );
+        });
+
+        CodeableConceptMappingUtils.extractTextOrCoding(observation.getInterpretation()).ifPresent(interpretation ->
+            narrativeStatementsBlock.append(
+                prepareNarrativeStatement(observation, interpretation, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper)
+            )
+        );
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java:210]:
```suggestion
        CodeableConceptMappingUtils.extractTextOrCoding(observation.getInterpretation())
            .map(interpretation -> prepareNarrativeStatement(observation, interpretation, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper))
            .ifPresent(narrativeStatementsBlock::append);
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java [hunk 4]
```diff
@@ -215,24 +161,85 @@ private String mapObservationToObservationStatement(IdMapper idMapper, Observati
...
+
+        CodeableConceptMappingUtils.extractTextOrCoding(observation.getBodySite()).ifPresent(bodySite -> {
+            String comment = BODY_SITE_PREFIX + bodySite;
+
+            narrativeStatementsBlock.append(
+                prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper)
+            );
+        });
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java:218]:
```suggestion
        CodeableConceptMappingUtils.extractTextOrCoding(observation.getBodySite())
            .map(BODY_SITE_PREFIX::concat)
            .map(comment -> prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper))
            .ifPresent(narrativeStatementsBlock::append);
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java [hunk 5]
```diff
@@ -215,24 +161,85 @@ private String mapObservationToObservationStatement(IdMapper idMapper, Observati
...
+
+        CodeableConceptMappingUtils.extractTextOrCoding(observation.getMethod()).ifPresent(method -> {
+            String comment = METHOD_PREFIX + method;
+
+            narrativeStatementsBlock.append(
+                    prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper)
+            );
+        });
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/ObservationMapper.java:226]:
```suggestion
        CodeableConceptMappingUtils.extractTextOrCoding(observation.getMethod())
            .map(METHOD_PREFIX::concat)
            .map(comment -> prepareNarrativeStatement(observation, comment, CommentType.LABORATORY_RESULT_DETAIL.getCode(), idMapper))
            .ifPresent(narrativeStatementsBlock::append);
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 6]
```diff
@@ -63,9 +63,12 @@ public String mapDiagnosticReportToCompoundStatement(DiagnosticReport diagnostic
```
Comment (matthewdoreilly) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:63]: No functional changes here. Minor refactoring changes for readability

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 7]
```diff
@@ -159,6 +132,38 @@ private Observation generateDefaultObservation(DiagnosticReport diagnosticReport
...
+                diagnosticReport, comment
+            );
+
+            reportLevelNarrativeStatements.append(narrativeStatementFromConclusion);
+        }
+
+        String codedDiagnosisText = diagnosticReport.getCodedDiagnosis().stream()
+            .map(CodeableConceptMappingUtils::extractTextOrCoding)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:151]:
```suggestion
            .flatMap(Optional::stream)
```

**Review by andrewwa-kainos — APPROVED** [review]
Just a few refactoring suggestions, to be more idiomatically Java 8+. Approved for if you prefer as is.

**Review by Broom94 — APPROVED** [review]

**Review by Broom94 — DISMISSED** [review]

File: service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_multiple_interpretations.xml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        <statusCode code="COMPLETE"/>
+        <availabilityTime value="20190601113000"/>
+    </NarrativeStatement>
+</component><component typeCode="COMP" contextConductionInd="true">
+    <NarrativeStatement classCode="OBS" moodCode="EVN">
+        <id root="5E496953-065B-41F2-9577-BE8F2FBD0757"/>
+        <text mediaType="text/x-h7uk-pmip">CommentType:LABORATORY RESULT DETAIL(E136)
+            CommentDate:20190601113000
+
+            Example without userSelected</text>
```
Comment (msolecka) [inline service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_multiple_interpretations.xml:45]: Shouldn't here be **Interpretation: Example without userSelected** ? isn't it from interpretation.coding[0].display ?

**Review by msolecka — CHANGES_REQUESTED** [review]

File: service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_interpretation_code_low.xml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        <statusCode code="COMPLETE"/>
+        <availabilityTime value="20190601113000"/>
+    </NarrativeStatement>
+</component><component typeCode="COMP" contextConductionInd="true">
+    <NarrativeStatement classCode="OBS" moodCode="EVN">
+        <id root="5E496953-065B-41F2-9577-BE8F2FBD0757"/>
+        <text mediaType="text/x-h7uk-pmip">CommentType:LABORATORY RESULT DETAIL(E136)
+            CommentDate:20190601113000
+
+            Low Value</text>
```
Comment (msolecka) [inline service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_interpretation_code_low.xml:45]: Interpretation: Low Value?

File: service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_interpretation_code_abnormal.xml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        <statusCode code="COMPLETE"/>
+        <availabilityTime value="20190601113000"/>
+    </NarrativeStatement>
+</component><component typeCode="COMP" contextConductionInd="true">
+    <NarrativeStatement classCode="OBS" moodCode="EVN">
+        <id root="5E496953-065B-41F2-9577-BE8F2FBD0757"/>
+        <text mediaType="text/x-h7uk-pmip">CommentType:LABORATORY RESULT DETAIL(E136)
+            CommentDate:20190601113000
+
+            Abnormal</text>
```
Comment (msolecka) [inline service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_interpretation_code_abnormal.xml:45]: Interpretation: Abnormal?

File: service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_multiple_interpretations.xml (hunk 8 above)
Comment (matthewdoreilly) [reply service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_multiple_interpretations.xml:45]: We do prepend "Interpretation: " for diagnosticReport.conclusion, but I don't think we should do it for observation.interpretation (from looking at the mapping spreadsheet).

See row 156 in DiagnosticReport
```
interpretation |
As per Observation (Uncategorised) but with the distinction that any interpretation that would previously have generated a value string for inclusion in pertinentInformation text should generate a NarrativeStatement containing the value text of comment type  LABORATORY RESULT DETAIL(E136) and date the same as the generated ObservationStatement/effectiveTime to be included within the CompoundStatement generated from the Observation
```

Doesn't say anything about prepending "Interpretation: ", unless you can see it somewhere else?

Comment (matthewdoreilly) [reply service/src/test/resources/ehr/mapper/diagnosticreport/observation/observation_compound_statement_with_multiple_interpretations.xml:45]: We do prepend "Interpretation: " for diagnosticReport.conclusion, but didn't realise we should do it for observation.interpretation as well. Can make that change now

**Review by msolecka — APPROVED** [review]

**Review by andrewwa-kainos — APPROVED** [review]

**Pull request merged**